config/main.tf

```diff
@@ -0,0 +1,39 @@
+terraform {
+  backend "s3" {
+    key = "terraform/build/callisto-auth-keycloak"
+  }
+}
+
+provider "keycloak" {
+  # Variables in here must be defined locally before running the plan
+  client_id     = var.tf_client_id
```

Comment:
Is this working I thought providers can't be set by variables?

Reply:
It's working for me, perhaps this became enabled in a recent version?

Reply:
ok I'm a little rusty. Reference values that are known before the configuration is applied are ok e.g. input variables are ok but attributes from other resource which can't be determined upfront are not